### DISCUSSION TITLE
HIP-412: Updating NFT JSON metadata schema standards

### DISCUSSION THREAD

**Discussion by Ashe-Oro** [hiero-ledger/hiero-improvement-proposals#413] [Edits]
A discussion around HIP412 and its updates to HIP10. 

https://github.com/hashgraph/hedera-improvement-proposal/pull/412

**Comment by web3-nomad** [comment]
re: localization, you may want/need to put in a file reference to which file is being localized. 
(in the example there are two URIs in the example NFT, and two localizations, but it is unclear which the localization is pointing too)

**Reply by rocketmay** [reply]
So, the localization in the example would point to different localized metadata that describe the root metadata, it doesn't actually point to the files themselves. The localized metadata would presumably point to localized files as well, as applicable.

**Reply by Ashe-Oro** [reply]
> 

inside the Files[] section, an individual file could have it's own localization metadata. In the example below, the first file "Example Video" doesn't contain localization data, but the second file [Example second file, a PDF" does. 

```json
{
    "name": "Example NFT",
    "creator": "Jane Doe, John Doe",
    "creatorDID": "did:hedera:mainnet:7Prd74ry1Uct87nZqL3ny7aR7Cg46JamVbJgk8azVgUm;hedera:mainnet:fid=0.0.123",
    "description": "This is an example NFT metadata",
    "format": "none",
    "files": [
        {
            "uri": "ipfs://bawlkjaklfjoiaefklankfldanmfoieiajfl",
            "type": "video/mp4",
            "metadata": {
                "name": "Example Video"
            }
        },
        {
            "uri": "ipfs://bawlkjaklfjoiaefklankfldanmfoieiajfl",
            "type": "application/pdf",
            "metadata": {
                "name": "Example second file, a PDF",
                "description": "The description is recommended but optional. The image provided is an optional preview",
                "image": "ipfs://bawlkjaklfjoiaefklankflda1313ieiajfl",
                "type": "image/jpeg",
                "localization": [    //per file localization    
                {
                    "uri": "ipfs://wgslkjaklfjoiaefklankflda132zafafewee",
                    "locale": "es"
                },
                {
                    "uri": "ipfs://ghasdgkjaklfjoiaefklankf12554wa6aga43ag",
                    "locale": "jp"
                }
                ]
            }
        }
    ]
```

The primary NFT (most likely the "Collection" would still have it's localization defined at the root. I've removed all files from the example here. 

```json
{
    "name": "Example NFT",
    "creator": "Jane Doe, John Doe",
    "creatorDID": "did:hedera:mainnet:7Prd74ry1Uct87nZqL3ny7aR7Cg46JamVbJgk8azVgUm;hedera:mainnet:fid=0.0.123",
    "description": "This is an example NFT metadata",
    "format": "none",
    "localization": [
        {
            "uri": "ipfs://bawlkjaklfjoiaefklankflda132zafga3tfa",
            "locale": "es"
        },
        {
            "uri": "ipfs://bawlkjaklfjoiaefklankf12554wa6aga4fda",
            "locale": "jp"
        }
    ]
}
```





Would this adequately cover all cases simply enough?

**Reply by robkainz** [reply]
Is there a default locale specified? If only Spanish and Japanese are supported, which one should we display by default if the user reads only English?

**Comment by Ashe-Oro** [comment]
<tokenID must be uint256 on Ethereum>

Per the ERC-721 specs on [ethereum.org](https://ethereum.org/en/developers/docs/standards/tokens/erc-721/), the tokenId for each NFT must be:

- uint256
- globally unique

Since tokenId's on Hedera will be in the format 0.0.xxxx (where xxxx is a unique number) will this pose an issue when bridging HTS NFTs to other chains?

**Reply by rocketmay** [reply]
Yes, Ashe and I briefly talked about this, but this actually seems like an issue. It's out of scope of NFT Metadata and more something to be addressed by the base platform. 

Right now the unique identifier for an NFT is the collection ID + the serial number. For example, 0.0.12345-1 would be a globally unique identifier. However it seems awkward/difficult to find a way to represent this in a UINT.

**Reply by Ashe-Oro** [reply]
the unique uint256 identifier for each NFT, in this example, seems to be 123451...where we drop the 0.0. and append the serial number '1'. 

While this would be unique on Hedera, it cannot be provable unique on ETH. I'm actually not sure how ERC721 NFT tokenId's are guaranteed unique, but I imagine they are assigned by the protocol. I'll research and report back.

**Reply by Ashe-Oro** [reply]
Ok, based on https://ethereum.stackexchange.com/questions/91292/how-to-set-tokenid-in-erc721-openzeppelin

....it seems our internal `tokenId` for NFTs should simply remove the dash and be used. That would ensure unique `tokenId` globally. On ETH, each contract assigns the `tokenId` and it must be unique within the ERC721 contract, but since the native Hedera protocol is assigning `tokenId` for each NFT, we already have globally unique IDs. 

Now...if our NFTs were to be bridged to a different chain (ETH, for example), we'd need to deploy an ERC721 like smart contract (or 1155, i'm not sure) that identifies `0.0.12345-1` as `123451`, for example, as the proper `tokenId`. Other chains will expect an actual ERC721 contract to manage NFTs with their associated actions(). 

Does this make sense @rocketmay @web3-nomad @Cooper-Kunz @paulatcalaxy 

If so, how can `tokenId` be included in the json metadata? Chicken and Egg problem, no?

**Reply by rocketmay** [reply]
The issue with removing the dash is that collection ID and serial ID are both arbitrary lengths, so there's no way to distinguish where the break is in a UINT.

(As discussed with you) I think that this is worth exploring in another discussion or another HIP, but for the purposes of getting this HIP out, this topic is strictly out of the scope of NFT Metadata since we wouldn't be storing the token ID/serial # in the metadata. 

As a thought, what if there was a SDK function which took the token-serial (0.0.12345-1) and run it through a hash function or similar process which produced a unique UINT, which could then be queried for? I'm sure there are plenty of performance/computational/technical considerations but could be a potential avenue for a solution.

**Comment by topshef** [comment]
is this a typo, should `localization` be within the `metadata` element?
![image](https://user-images.githubusercontent.com/40291610/166612998-c05b5322-6050-42a2-94a1-578d91a48b21.png)

**Reply by rocketmay** [reply]
Yes, this is a typo.

**Comment by robkainz** [comment]
HIP 412 states "For resources that are on the world wide web, the standard http and https protocols are acceptable", however it is not acceptable to Apple by default for mobile apps built using Xcode (nor should it be IMHO) and therefore security must be reduced globally for the mobile app to allow any resource to be accessed via http. I would recommend only supporting https for this reason, to prevent potentially mistakenly-coded http calls for sensitive data that should be using secure communication. Can the requirement to support the insecure http protocol be removed? Or must we move forward without a wise safeguard in our mobile financial apps that need to query NFT data that could result in insecure communications elsewhere in the app should a mistake be made? It's like disabling an important firewall rule to make NFT creators happy that don't use secure urls for their metadata or resources.

**Reply by gregscullard** [reply]
I would tend to agree that https is preferable in almost all cases to http. Maybe the specification should mandate use of https and suggest http should only be used if/when there is no alternative.

**Reply by robkainz** [reply]
Unfortunately if we need to support http at all, I have to lower the shield. But I suppose a properly implemented countermeasure can be put into place on my part to ensure no Wallawallet developer inadvertently uses http to interact with other 3rd-party services we may depend on in the future.

**Reply by rocketmay** [reply]
Agreed. Let's slot removing HTTP from the specification in as a change for the next version.

**Comment by nigelg71** [comment]
I can see this feature would be useful for us at TYMLEZ. As the properties can be any JSON objects, they could be used to provide additional information about the token without the need to retrieve the full VP e.g. for a Carbon Removal Unit (CRU) token, we could provide information about the project and the nature of the removals, or for a Carbon Emissions Token (CET), we could include the green house gas breakdown. This would make it easier to exchange tokens via a marketplace.

Are there any thoughts about how this would be incorporated into the Guardian? I think we would need the guardian to allow us to specify a scheme for the properties, with the values able to be set in the policy before minting. It may be worth considering using the files property to link to the VP instead of navigating from the MEMO field.

Also, what are the likely cost implications for token minting? I assume the cost would increase based on the size of the metadata (much like it does for the MEMO field).

**Comment by mattsmithies** [comment]
I believe that the HIP412 spec should be stricter in the JSON definition, particularly around the optional array of objects, like **files** and **localization**, I would assume that we would require those particular sub-fields.

For example, for **localization** sub-objects there are **uri** and **locale** which I suspect are hard requirements, I might go as far as to say that those are the only properties that are needed.

Furthermore, perhaps it would be a good idea to integrate the validation of the types we are seeking within the JSON schema so we can be stricter for all downstream clients. Using [schema.org](https://schema.org/) as a base.

This snippet could be an example of the updated JSON schema for NFT spec conformance. 

@rocketmay @gregscullard, is this something you could buy into, would it be useful?

```
"localization": {
    "type": "object",
    "description": "Array of localized metadata.",
    "properties": {
        "uri": {
            "type": "string",
            "description": "A URI pointing to the localized metadata json file for this locale.",
            "$ref": "https://schema.org/url"
        },
        "locale": {
            "type": "string",
            "description": "A two-letter locale.",
            "$ref": "https://schema.org/addressCountry"
        }
     "required": [ "uri", "locale" ]
     "additionalProperties": false
}             
```

**Reply by rocketmay** [reply]
I have no strong opinions on this. Stricter JSON definition to adhere to existing standards sounds like a good plan. Localization is not anywhere near my area of knowledge, so if you say it should be formatted this way and the community agrees then this is a fine suggestion.

**Comment by michielmulders** [comment]
Hi all, I took some time to review the NFT spec and wanted to add some thoughts/questions. Let me know what you think?

### Suggestions

1. Add a `preview_image` field to quickly load an image rather than loading it from IPFS. 
**Motivation:** NFT explorers, wallets, and related tooling will first load the image from IPFS, which can take a while if it hasn't been cached on their servers. Therefore, allow creators to upload an optimized version of their NFT hosted at a traditional cloud provider.

2. Include an `image_integrity` field to add a sha256 of the NFT. This can be useful for the original work but also for the `preview_image` to ensure the preview_image doesn't get swapped for another image because it's hosted at a cloud provider.

3. Make a stronger specification for defining "rarity properties" of NFT collections. It will make it easier for NFT tooling to calculate rarity scores and display their properties. OpenSea and other chains implement an `attributes` field with the following values.

- `trait_type`: string
- `value`: string (Default: "None")
- `max_value`: string -> optional (if set, it indicates that you can parse the `value` field as an integer - created by OpenSea)

Examples can be found here: [MetaPlex standard](https://docs.metaplex.com/programs/token-metadata/token-standard#the-non-fungible-standard) and [Algorand ARC0003 standard](https://algoexplorer.io/tx/FMK7DH6IXKSSWVDO3RS6J3MKJ7OXQCOTR4EDWKJQOGLRS6BR3MEQ). 

4. Add a field that indicates the token standard used. Standards evolve, so it would be useful to indicate the standard in a separate field. This makes it easier for NFT tooling to detect how to parse the data. e.g. `@HIP412`

5. Add an `external_url` field that links to the project's main website/social. 

### Thoughts

1.  What are the possible use cases for indefinitely nested files?

2. Do we support fractionalized NFTs? Like owning a share of a piece of art represented by an NFT? It can be achieved by including two more properties in a standard:

- `units`: Must be a power of 10: 1, 10, 100, 1000, ...
- `decimals`: Must be equal to the logarithm in base 10 of the total number of `units`

e.g. 10 `units` and 1 `decimals` equals 10 fractional shares

**Reply by Ashe-Oro** [reply]
Great feedback, thank you. I'll address them numerically. 

1. `preview_image` would be very helpful and I agree that it should be added. We had originally had the idea of a preview image for non-image based NFTs which would use the image field, but having a dedicated `preview_image` value is cleaner and doesn't overload. 

2. `image_integrity` seems like a good solution to prevent phishing introduced by adding `preview_image`.  

3. `max_value` is [referenced ](https://github.com/hashgraph/hedera-improvement-proposal/blob/a902a68f40f8ca0b10cab3498414e69a3385e1b3/HIP/hip-412.md?plain=1#L326 )in the original spec, but only as an option field in the OpenSea standard.  Perhaps it's not specifically called out to the point that people will know it's available? 

4. Fair point. We can add a field for `token-standard` or `schema-standard`. @rocketmay thoughts? 

5. While this info can be included in the properties array, I have seen it often included in the root. I think we should add `external_url` to the root since it's commonly used there. That said, all social links (etc) should be included in `properties`.

**Reply by michielmulders** [reply]
RE 3: I would indeed call it out more specifically - it's important to get right from the get-go to avoid differences in trait definitions (also set the default value for traits) to make it easier for tools like NFT explorers/wallets to pick up this information. 

RE 5: I would vouch for an `external_url` property in the root. Other social links can indeed be added to the `properties`, but it's just a minor suggestion.

Also, what do you think about `Thoughts -> 2`? Having a standard that also works for fractional NFTs would be great to address even more NFT use cases. If we don't address this, the use case will arise sooner than later.

**Reply by HGraphPunks** [reply]
Hey, cool ideas! 

Here are some of my thoughts
Suggestions

1. Add a preview_image field to quickly load an image rather than loading it from IPFS.

This is already accounted for in the current standard. (edit: not specifically from IPFS loading issues, but from file size issues. I'd also say, I won't ever use a URL in a NFT that isn't IPFS. The preview image would at best be used by some to link to a HTTP server until that server url structure breaks eventually in 1-2-5-10 years)  

A great example is the token I just minted for Hangry Baboons. The 3d coin COULD be the main image, but as set by the standard 412, the image value should be easily loadable by wallets, markets, etc. So the main image on this 3d item is instead a much smaller, easily presentable in wallets and markets png, with the files array containing more information / images / and files that can pertain to the main NFT purpose:
<img width="372" alt="Screen Shot 2022-08-24 at 4 14 55 PM" src="https://user-images.githubusercontent.com/91838723/186533075-418ac038-aa1f-435c-8772-024856097ee9.png">

`image_integrity` IMHO could be present in a format:"web url" or some other delineation. Most use IPFS and should be immutable data from that CID


2. Make a stronger specification for defining "rarity properties" of NFT collections. It will make it easier for NFT tooling to calculate rarity scores and display their properties. OpenSea and other chains implement an attributes field with the following values.

trait_type: string
value: string (Default: "None")
max_value: string -> optional (if set, it indicates that you can parse the value field as an integer - created by OpenSea)
Examples can be found here: MetaPlex standard and Algorand ARC0003 standard.

I personally don't see the benefit to the max_value in the attributes array as string sensing will still have to be done either way.  Even when set, code can't depend on the data in max_value to accurately determine if you can parse value. 

4. Add a field that indicates the token standard used. Standards evolve, so it would be useful to indicate the standard in a separate field. This makes it easier for NFT tooling to detect how to parse the data. e.g. @HIP412

I agree some delineation of standard used in the metadata could help future reading of the data to know where to expect certain things. 


5. Add an external_url field that links to the project's main website/social.
I don't think external_url is terribly helpful on the root. I think it being available in the properties array, or if a new format is proposed with it on the root that could work, but as the default I think it'll lead to a lot of dead links and non usage.

**Reply by Ashe-Oro** [reply]
`external_url` is included on the root for OpenSea and I think we should follow their lead https://docs.opensea.io/docs/metadata-standards

**Reply by HGraphPunks** [reply]
Open Sea is a great standard to look at for inspiration but not 100% replicate it.
Though external_url  is. present, is it used?
Checking Bored Ape, Mutant Ape, Azuki contracts, none have that filled in. Still not sure it's useful.

**Reply by rocketmay** [reply]
1. Preview image

This is actually one of the discussion points right now in the community. The current thought right now is to add an optional new key/value pair: "isThumbnail":"True", which can be added to any file (which must be an image) within the "files" array. This would be in concert with another optional "isDefaultFile":"True" pair which would specify that a particular file in the file array is the main file. Programs which can render the default file may choose to render that instead of a thumbnail.

2. Image_integrity

Agree that this has value. Propose naming this to "sha256_checksum" for clarity. The main image and the files in the file array can have this specified as an optional key.

3. Specification for NFT properties

I agree that a standardized art/collectible NFT schema would be valuable. HIP412 is intended to be a base specification across all use cases of NFT's not just open sea standard NFTs. I propose we start a new HIP which extends HIP412, to define these properties.

4. Token Standard field

The "Format" field specified in HIP412 is intended to serve this purpose. Tokens are assumed to follow the HIP412 standard, unless Format indicates otherwise. And even then HIP412 is intended to be the base standard for other schemas to be built on top of.

5. External_url field

Propose this could be part of the collectible NFT HIP mentioned in my comment for "3". Many NFTs have no use for External_url, so I hesitate to bloat the HIP412 specification with this kind of information. As a side note, since external_url is specified in the open sea specification, theoretically NFTs with the external_url field should be covered since the Open Sea format is implicitly supported in the standard if "Format" is not defined.

**Reply by rocketmay** [reply]
Comment on the Additional Thoughts:

Infinite-nested NFT: It's not so much about the ability to nest NFT's infinitely but about recursively being able to process the files in the metadata. If each sub-file follows the same formatting (which implicitly allows nesting), the code which processes the metadata can be far simpler.

Fractionalized NFTs:

Propose an additional HIP to cover this. In which case I suppose "Format" might be modified to be a string of comma-separated tags so you could have a "format":"open-sea, fractional", but maybe that starts to get silly.

The main thing here is fractionalized NFTs are a valid use case, but their implementation should not be concerned inside of the HIP412 standard that broadly addresses all NFTs.

**Reply by michielmulders** [reply]
@rocketmay 

1. Preview image: I like this idea of setting an extra property to say if it's default/preload image. It's a small change that solves the issue.

3. Creating a new HIP that ties into NFTs would be great to make it easier for NFT tooling to interpret metadata. I think having this "Token Standard Field" will become more critical. You can at least specify the `standard@specification` you are using. 

X. Fractionalized NFTs: Perhaps another extension to address fractionalized NFT use cases? It can be pretty simple with just two properties `units` and `decimals`.

**Reply by hashaxis** [reply]
@michielmulders Just a note on point 3 here:

Incase you were unaware, currently many projects are already using 'attributes' to define traits and then 'rarity' is deduced from that based on the percentage of tokens within a TokenID possessing specific properties.

**This method is already wide spread across the majority of major projects that are using traits at all.**

```
"attributes": [
        {
            "trait_type": "coolness",
            "value": "50",
            "max-value": "100",
        },
        {
            "trait_type": "colour",
            "value": "red"
        },
    ],

```

Example NFT from a popular collection:
https://cloudflare-ipfs.com/ipfs/bafyreiemzhyltoedjwlkcvsxlgglwxxrucvs7vmz6fuvuxpgvg752atn4y/metadata.json

**Reply by michielmulders** [reply]
Draft for an extension HIP for Web3 NFTs has been created here as a PR: https://github.com/hashgraph/hedera-improvement-proposal/pull/565 - Any feedback is welcome!

**Reply by justynspooner** [reply]
Great points to raise @michielmulders 

There is a recently submitted HIP covering Fractionalised NFTs [here](https://github.com/hashgraph/hedera-improvement-proposal/pull/543).

> 1. Add a preview_image field to quickly load an image rather than loading it from IPFS

I think @HGraphPunks and @rocketmay cover this well already. I like the idea of optional fields at the file level such as  `is_default_file` and not introducing another top level field.

> 2. Include an image_integrity field to add a sha256 of the NFT

If files are on IPFS then there's no need for this. If we feel this is useful, then @rocketmay suggestion naming this optional field as `sha256_checksum` sounds good.

> 3. Make a stronger specification for defining "rarity properties" of NFT collections

Agree with @hashaxis. The opensea attributes standard works really well for this and is already a widely adopted standard. Would be good to formalise that as a standard way of defining rarity traits rather than forcing the format to be "opensea" to use them as is the case with 412 currently.

> 4. Add a field that indicates the token standard used

As already mentioned, `format` is designed for this although I think we could go one step further and use existing standards to formalise this.

Here is an examples of how Microsoft express the format/schema in their JSON files: 
https://github.com/Azure/azure-quickstart-templates/blob/master/100-blank-template/metadata.json

```
{
  "$schema": "https://aka.ms/azure-quickstart-templates-metadata-schema#",
  "type": "QuickStart",
  "itemDisplayName": "Blank Template",
  "description": "A blank template and empty parameters file.",
  "summary": "A blank template and empty parameters file.  Use this template as the framework for your custom deployment.",
  "githubUsername": "bmoore-msft",
  "dateUpdated": "2022-07-05"
}
```

The `$schema` is a URL to the schema definition which is published separately: https://aka.ms/azure-quickstart-templates-metadata-schema#

We're half way there with the `format` field, all we need to do is rename to `$schema`, publish the schema definition and link to it. The schema could be hosted on IPFS or some publicly accessible schema repository.

An example metadata that says it follows the standard base HIP 412 schema:
```
{
    "$schema": "ipfs://<CID>/hip412.schema.json",
    "name": "Token name",
    "creator": "Artist name",
    "image": "ipfs://<CID>/image.jpg"
    …etc
}
```

If someone wants to extend the base schema, then they can just create their own schema or link to an existing one that extends the base `hip412.schema.json` then put a link to their custom schema in the `$schema` field. As long as they are using HIP 412 schema as the base schema then they can be guaranteed that the NFTs base data will be displayed correctly.

For example, here is a custom JSON Schema that inherits all the properties of HIP 412 but designed specifically to describe the properties of a game character.

This schema could be published to `ipfs://<CID>/hip412GameCharacter.schema.json` for example
```
{
  // Use all the properties of the base HIP 412 schema
  "allOf": [{ "$ref": "ipfs://<CID>/hip412.schema.json" }],
  // Now describe the extended properties we want to add
  "properties": {
    "character_attributes": {
      "type": "object",
      "properties": {
        "strength": { "type": "number" },
        "speed": { "type": "number" },
        "intelligence": { "type": "number" }
      }
    }
  }
}
```

Then the actual metadata could be:
```
{
    "$schema": "ipfs://<CID>/hip412GameCharacter.schema.json",
    "name": "Token name",
    "creator": "Artist name",
    …etc
    "character_attributes": {
        "strength": 5,
        "speed": 8,
        "intelligence": 9
    }
}
```

NFT listing services such as marketplaces/wallets could then decide to just support the base HIP 412 schema or they could start opting in to start supporting other custom schemas that extend HIP 412 that become popular in the community.

The other great thing about having a published schema definition that is referenced in the JSON metadata is that we can write automatic JSON schema validators / parsers in our minting tools and NFT listing services using libraries such as [AJV](https://ajv.js.org/).

**Reply by michielmulders** [reply]
@justynspooner I really like the schema idea and how you can define your own JSON schema. Instead of naming it `$schema` (which is the default outside of blockchain), I'm seeing a lot of use for the `format` property.

**Reply by justynspooner** [reply]
It doesn't really matter what we name it, the main thing is that it links to a JSON Schema file which describes exactly what the structure is.

**Comment by topshef** [comment]
Maybe worth casting an eye on the current LG format, and connecting with them.. it seems slightly off from HIP-412
https://gomint.me/explore/NFT/?network=mainnet&tokenId=0.0.1182876-1

**Reply by michielmulders** [reply]
Thanks for your reply. I see they have defined the following properties which can go into the `properties` array.

```
"edition_no": 1,
"total_of_editions": 350,
```

Not sure why they need width and height for the files array? But other than that, the standard defined in the PR is very similar to what they have done.

**Reply by justynspooner** [reply]
They also need to make sure the `files` array is it at the top level if following HIP 412 or define format as `opensea` in order to have the files array in the properties object like it currently is.

The metadata is also stored centrally at https://api.lgartlab.com/api/v1/meta/e215489f56d6ed3dbac92e26f72a643d so could be changed at any time. Ideally it should be stored on IPFS.

**Reply by michielmulders** [reply]
I would argue to follow opensea in this case, put it in the `properties` object. Unless there's a clear reason why not to do it (e.g. follow our base HIP 412 standard)?

**Comment by jaycoolslm** [comment]
Very late to the party but have just noticed that 'creator / creatorDID' is not referenced in the json schema defnition

https://deploy-preview-412--hedera-hips.netlify.app/hip/hip-412
